pkg/neonvm/controllers/vmmigration_controller.go

```diff
@@ -150,25 +150,35 @@ func (r *VirtualMachineMigrationReconciler) Reconcile(ctx context.Context, req c
 		return ctrl.Result{}, err
 	}
 
-	getVM := func() (*vmv1.VirtualMachine, error) {
-		var vm vmv1.VirtualMachine
-		err := r.Get(ctx, types.NamespacedName{Name: migration.Spec.VmName, Namespace: migration.Namespace}, &vm)
-		if err != nil {
-			log.Error(err, "Failed to get VM", "VmName", migration.Spec.VmName)
-			return nil, err
+	// Fetch the corresponding VirtualMachine instance
+	var vmObject vmv1.VirtualMachine
+	err := r.Get(ctx, types.NamespacedName{Name: migration.Spec.VmName, Namespace: migration.Namespace}, &vmObject)
+	vm := &vmObject
+	if err != nil {
+		log.Error(err, "Failed to get VM", "VmName", migration.Spec.VmName)
+		if apierrors.IsNotFound(err) {
+			// stop reconcile loop if vm not found (already deleted?)
+			message := fmt.Sprintf("VM (%s) not found", migration.Spec.VmName)
+			r.Recorder.Event(migration, "Warning", "Failed", message)
+			meta.SetStatusCondition(&migration.Status.Conditions,
+				metav1.Condition{
+					Type:    typeDegradedVirtualMachineMigration,
+					Status:  metav1.ConditionTrue,
+					Reason:  "Reconciling",
+					Message: message,
+				})
+			migration.Status.Phase = vmv1.VmmFailed
+			return r.updateMigrationStatus(ctx, migration)
 		}
-		return &vm, nil
+		// return err and try reconcile again
+		return ctrl.Result{}, err
```

Comment:
question(blocking): Given that `if err != nil` always ends in returning here, does that mean that we never remove the finalizer, and so the migration object cannot be deleted if the VM doesn't exist?

Reply:
That's true, and this problem existed before this PR. Let me think about a solution.